### PR TITLE
Support for multiple optimizer groups in torch 2.0 + FSDP

### PR DESCRIPTION
**Pull request by sashaDoubov**
# What does this PR do?

This add support for multiple optimizer param groups when using FSDP + `use_orig_params=True` in torch 2.0. 

# What issue(s) does this change relate to?

This is an unsupported feature that is blocking for features like muP parameterization or large-scale training with different learning rates per layer.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/dist_strategy.py [hunk 1]
```diff
@@ -123,6 +123,49 @@ def prepare_ddp_module(module: torch.nn.Module, find_unused_parameters: bool) ->
                        'with distributed support.')
 
 
+def _recreate_fsdp_param_groups_from_unwrapped_opt_info(fsdp_wrapped_named_params: Iterator[Tuple[str,
```
Comment (sashaDoubov) [inline composer/trainer/dist_strategy.py:126]: some of these names may be a little long so wanted to check with reviewers whether this is acceptable formatting

**Review by dakinggg — COMMENTED** [review]
LGTM, but would like @bcui19 to approve

File: composer/trainer/dist_strategy.py [hunk 2]
```diff
@@ -123,6 +123,49 @@ def prepare_ddp_module(module: torch.nn.Module, find_unused_parameters: bool) ->
...
+
+    for fsdp_name, param in fsdp_wrapped_named_params:
+
+        unwrapped_name = clean_tensor_name(fsdp_name)
+        # need to have a 1:1 mapping between a fsdp param name and the non-wrapped vanilla param name
+        retrieved_group_num = non_wrapped_param_names_to_group_num[unwrapped_name]
+        group_num_to_optimizer_info[retrieved_group_num]['params'].append(param)
+
+    # the values of the list contains the list of param groups
+    return list(group_num_to_optimizer_info.values())
```
[the remark this answers is not in the document]
Comment (sashaDoubov) [reply composer/trainer/dist_strategy.py:166]: It's probably best to preserve the order of the param groups, between unwrapped <-> wrapped params so I'll address this

File: tests/trainer/test_fsdp_param_groups.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    num_classes = 10
+    model = SimpleModel(num_features=1, num_classes=num_classes)
+    dataset = RandomClassificationDataset(shape=(num_classes,), size=2, num_classes=num_classes)
+    dataloader = DataLoader(dataset, sampler=dist.get_sampler(dataset))
+
+    # create a different parameter per group
+    param_groups = [{'params': param, 'lr': (0.1 + 0.1 * i)} for i, param in enumerate(model.parameters())]
+    optimizer = torch.optim.SGD(param_groups, lr=0)
+
+    with pytest.raises(RuntimeError):
```
[the remark this answers is not in the document]
Comment (sashaDoubov) [reply tests/trainer/test_fsdp_param_groups.py]: updated!

File: composer/trainer/dist_strategy.py [hunk 4]
```diff
@@ -123,6 +123,49 @@ def prepare_ddp_module(module: torch.nn.Module, find_unused_parameters: bool) ->
                        'with distributed support.')
 
 
+def _recreate_fsdp_param_groups_from_unwrapped_opt_info(fsdp_wrapped_named_params: Iterator[Tuple[str,
+                                                                                                  torch.nn.Parameter]],
+                                                        non_wrapped_param_names_to_group_num: Dict[str, int],
+                                                        group_num_to_optimizer_info: Dict[int, Dict[str, Any]]):
```
[the remark this answers is not in the document]
Comment (sashaDoubov) [reply composer/trainer/dist_strategy.py]: updated with return type

File: composer/trainer/dist_strategy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sashaDoubov) [reply composer/trainer/dist_strategy.py]: done

File: tests/trainer/test_fsdp_param_groups.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import torch
+from packaging import version
+from torch.utils.data import DataLoader
+
+from composer.trainer.trainer import Trainer
+from composer.utils import dist, misc
+from tests.common import RandomClassificationDataset, SimpleModel, device, world_size
+
+
+@pytest.mark.parametrize('mixed_precision', ['FULL', 'DEFAULT', 'PURE'])
```
[the remark this answers is not in the document]
Comment (sashaDoubov) [reply tests/trainer/test_fsdp_param_groups.py]: I have changed the test to only test for `'DEFAULT'`

**Review by sashaDoubov — COMMENTED** [review]
Thanks for the review @dakinggg, I have updated the PR to address your comments.

File: tests/trainer/test_fsdp_param_groups.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+from composer.trainer.trainer import Trainer
+from composer.utils import dist, misc
+from tests.common import RandomClassificationDataset, SimpleModel, device, world_size
+
+
+@pytest.mark.parametrize('mixed_precision', ['DEFAULT'])
+@pytest.mark.parametrize('reentrant', [True, False])
+@pytest.mark.filterwarnings('ignore::UserWarning')
+@device('gpu')
+@world_size(1, 2)
```
Comment (bcui19) [inline tests/trainer/test_fsdp_param_groups.py:21]: For FSDP tests, I think we should really just test with `world_size=1`?

File: tests/trainer/test_fsdp_param_groups.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    max_duration='3ba',
+                    device=device)
+    gc.collect()
+
+
+@pytest.mark.parametrize('mixed_precision', ['FULL', 'DEFAULT', 'PURE'])
+@pytest.mark.parametrize('reentrant', [True, False])
+@pytest.mark.filterwarnings('ignore::UserWarning')
+@device('gpu')
+@world_size(1, 2)
```
Comment (bcui19) [inline tests/trainer/test_fsdp_param_groups.py:61]: Same as above

**Review by bcui19 — APPROVED** [review]
this is great! Mostly LGTM, I think we shouldn't be testing FSDP w/ world size=1

File: tests/trainer/test_fsdp_param_groups.py (hunk 6 above)
Comment (sashaDoubov) [reply tests/trainer/test_fsdp_param_groups.py]: changed to run on world size 2 only!

File: tests/trainer/test_fsdp_param_groups.py (hunk 7 above)
Comment (sashaDoubov) [reply tests/trainer/test_fsdp_param_groups.py]: fixed

**Pull request merged**